Introduce New Routing Mode Options [API-2240] (2/4) (#901)
hazelcast/hazelcast-csharp-client: doc/dev/doc/network.md

By default, the provided list is shuffled and tried in a random order. You can disable this behaviour by setting the configuration option `hazelcast.networking.shuffleAddresses` to `false`. In this case the address list will be tried in the specified order. 

## Operation Mode
The client has three operation modes because of the distributed nature of the data and cluster: all members, multi members and single member. All members routing is enabled by default, and is controlled by the `@Hazelcast.NetworkingOptions.RoutingMode` configuration option.

> [!NOTE]

> > Smart Routing is deprecated and will be removed in the next major release. Please use the new RoutingMode mode instead.

Reviewer: When I read that it sounded like the semantics was deprecated -- this is in reference to `hazelcast.networking.smartRouting` and not the smart routing semantics, that's just been renamed. Maybe clarify it a bit?

Author: I see, you are right. I've paraphrased it. How about now?

### Routing Mode

Routing mode allows you to choose the routing strategy of the client. 

There are three routing modes:

> * `@Hazelcast.Networking.RoutingModes.SingleMember` aka Unisocket mode.
> * `@Hazelcast.Networking.RoutingModes.MultiMember` connects to subset of the cluster.
> * `@Hazelcast.Networking.RoutingModes.AllMembers` connects to all members in the cluster aka Smart mode.

Reviewer 2: ```suggestion
* `@Hazelcast.Networking.RoutingModes.SingleMember` connects to a single member of the cluster (previously Unisocket routing).
* `@Hazelcast.Networking.RoutingModes.MultiMember` connects to subset of members in the cluster.
* `@Hazelcast.Networking.RoutingModes.AllMembers` connects to all members of the cluster (previously Smart routing).
```

Author: Fixed.

#### Single Member Routing Mode

For some cases, the clients can be required to connect to a single member instead of each member in the cluster. Firewalls, security or some custom networking issues can be the reason for these cases.

In the unisocket client mode, the client will only connect to one of the configured addresses. This single member will behave as a gateway to the other members. For any operation requested from the client, it will redirect the request to the relevant member and return the response back to the client returned from this member.
#### Multi Member Routing Mode
Multi member routing mode is a mode where the client connects to a subset of the cluster members. This mode behaves based
on given strategy on `@Hazecast.Networking.NetworkingOptions.RoutingMode.Strategy`.

Available strategies are:
* `@Hazelcast.Networking.RoutingStrategy.PartitionGroups` which is the default strategy. Clients connects to subset of the cluster based on partition groups which contains the current connected member. 

> Please note that server should support this routing mode which is available with Hazelcast 5.5.0.

Reviewer 2: ```suggestion
Please note that server should support this routing mode which is available with Hazelcast Enterprise 5.5.0.
```

Author: Fixed.

#### All Members Routing Mode

> In all members mode, the clients connect to each cluster member. Since each data partition uses the well known and consistent hashing algorithm, each client can send an operation to the relevant cluster member, which increases the overall throughput and efficiency. Smart mode is the default mode.

Reviewer 2: ```suggestion
In all members mode, the clients connect to each cluster member. Since each data partition uses the well known and consistent hashing algorithm, each client can send an operation to the relevant cluster member, which increases the overall throughput and efficiency. This mode is the default mode.
```

Author: Fixed.
